mm2src/coins/eth/wallet_connect.rs

```diff
@@ -0,0 +1,253 @@
+            "nonce": u256_to_hex(self.nonce),
+            "from": format!("0x{}", hex::encode(self.my_address.as_bytes())),
+            "gas": u256_to_hex(self.gas),
+            "value": u256_to_hex(self.value),
+            "data": format!("0x{}", hex::encode(self.data))
+        });
+
+        if let Some(gas_price) = self.gas_price {
+            tx_json
+                .as_object_mut()
+                .unwrap()
+                .insert("gasPrice".to_string(), json!(u256_to_hex(gas_price)));
+        }
+
+        let to_addr = match self.action {
+            Action::Create => None,
+            Action::Call(addr) => Some(addr),
+        };
+        if let Some(to) = to_addr {
+            tx_json
+                .as_object_mut()
+                .unwrap()
+                .insert("to".to_string(), json!(format!("0x{}", hex::encode(to.as_bytes()))));
+        }
+
+        Ok(json!(vec![tx_json]))
+    }
+}
+
+#[async_trait::async_trait]
+impl WalletConnectOps for EthCoin {
+    type Error = MmError<EthWalletConnectError>;
+    type SignTxData = (SignedTransaction, BytesJson);
+    type SendTxData = (SignedTransaction, BytesJson);
+    type Params<'a> = WcEthTxParams<'a>;
+
+    async fn wc_chain_id(&self, wc: &WalletConnectCtx) -> Result<WcChainId, Self::Error> {
+        let chain_id = WcChainId::new_eip155(self.chain_id.to_string());
+        wc.validate_update_active_chain_id(&chain_id).await?;
+
+        Ok(chain_id)
+    }
+
+    async fn wc_sign_tx<'a>(
+        &self,
+        wc: &WalletConnectCtx,
+        params: Self::Params<'a>,
+    ) -> Result<Self::SignTxData, Self::Error> {
+        let bytes = {
+            let chain_id = self.wc_chain_id(wc).await?;
+            let tx_json = params.prepare_wc_tx_format()?;
+            let tx_hex: String = wc
+                .send_session_request_and_wait(&chain_id, WcRequestMethods::EthSignTransaction, tx_json, Ok)
+                .await?;
+            // First 4 bytes from WalletConnect represents Protoc info
+            hex::decode(&tx_hex[4..])?
+        };
+        let unverified = rlp::decode(&bytes)?;
+        let signed = SignedTransaction::new(unverified)?;
+        let bytes = rlp::encode(&signed);
+
+        Ok((signed, BytesJson::from(bytes.to_vec())))
+    }
+
+    async fn wc_send_tx<'a>(
+        &self,
+        wc: &WalletConnectCtx,
+        params: Self::Params<'a>,
+    ) -> Result<Self::SignTxData, Self::Error> {
+        let tx_hash: String = {
+            let chain_id = self.wc_chain_id(wc).await?;
+            let tx_json = params.prepare_wc_tx_format()?;
+            wc.send_session_request_and_wait(&chain_id, WcRequestMethods::EthSendTransaction, tx_json, Ok)
+                .await?
+        };
+        let tx_hash = tx_hash.strip_prefix("0x").unwrap_or(&tx_hash);
+        let maybe_signed_tx = {
+            // Wait for 10 seconds for the transaction to appear on the RPC node.
+            let wait_rpc_timeout = 10_000;
```

Comment:
not very quick for 10s timeout really. I thought about it and decided to increase this timeout to 30secs for WC since some txs might actually take longer than 10s to appear on chain.. still depends on the WC wallet used. So I think 30 secs is a good timeout. 

Metasmask impl uses 60secs timeout also.

Reply:
done